### PR TITLE
Change item model loading priority to fix model conflicts

### LINKED ISSUE #3182: Blockstate and item json with same name causes strange issues
Blockstate (blockstates/root.json:

```
{
  "variants": {
    "normal": { "model": "botania:root"  } // should point to models/block/root.json
  }
}
```

Item (models/item/root.json) :

```
{
    "parent": "minecraft:item/generated",
    "textures": 
    {
        "layer0": "botania:items/root"
    }
}
```

These two simply happened to be named the same. The former is for a block called "botania:root" the latter is simply a meta variant of "botania:manaResource" mapped to a custom name.

What happens is when I do `ModelLoader.setCustomModelResourceLocation(Item.getItemFromBlock(myBlock), 0, "botania:root#normal")`, not only does the block's item not take the model specifies `models/block/root.json`, it also _back propagates the item/generated model onto the block model in-world_. That's extremely extremely extremely strange behaviour and I doubt it's intended.

### PR DESCRIPTION
**Pull request by bs2609**
Fixes #3182.

Changes `ModelLoader` to try loading a blockstate variant first, then the separate item model.

This is needed to prevent conflicts - as the item model location doesn't encode any variant information, two objects can share the same "candidate" item model location. Because loading from this location is prioritised at present, this means that, if the item model exists, both things will be mapped to it. Trying the more specific variant location first fixes this issue.

As a consequence, the handling of multipart models has been altered to be a bit stricter. Now a multipart model will only be created for variants that are registered by `registerMultipartVariant` (as opposed to the existing check here of just `definition.hasMultipartData()`). Without this change, the priority switch causes items corresponding to blocks using multipart models to be assigned the multipart model, instead of falling through to the item location.

A test mod is included, which has a setup similar to that described in the issue.

### CODE REVIEW

File: src/main/java/net/minecraftforge/client/model/ModelLoader.java [hunk 1]
```diff
@@ -24,9 +24,7 @@
 import java.awt.image.BufferedImage;
 import java.util.ArrayList;
 import java.util.Collection;
-import java.util.Collections;
```
Comment (ChloeDawn) [inline src/main/java/net/minecraftforge/client/model/ModelLoader.java]: All of these import cleanups aren't really necessary for this PR, could probably amend it and revert the changes to them

**Comment by josephcsible** [comment]
This change broke existing mods: McJtyMods/XNet#288

**Comment by BlayTheNinth** [comment]
This appears to have also broken TESR items, unless I've been doing it wrong:

[Item Model JSON](https://github.com/blay09/Waystones/blob/1.12/src/main/resources/assets/waystones/models/item/waystone.json)

[Model Registration](https://github.com/blay09/Waystones/blob/ea6189141f19e61cfab0892875308ae866174b24/src/main/java/net/blay09/mods/waystones/Waystones.java#L114-L119)

https://github.com/blay09/Waystones/issues/126

With Forge 2715 the item renders fine, in 2716 it's invisible. I know TESR items are discouraged and I plan on porting it to JSON once I have time during the 1.13 update, but this was previously functional. Is there any simple short-term solution I can apply for this?

The model loaded now returns `false` for `isBuiltInRenderer` and thus the TESR rendering path never gets called.

**Comment by bs2609** [comment]
It seems that most of the issues here are caused by people having specified both possible item model definitions and then relying on a certain one being chosen. Perhaps the model loading code should warn about duplicate definitions?

**Comment by Glease** [comment]
It seems a fix is pretty straight forward. That PR made blockstate #inventory variant to take precedency instead of item model json. As suggested in that PR removing the duplicate definition of a model will resolve the issue. In this case removing the #inventory variant from blockstate json make everything fine.
I've tested my solution both on 2747 (a version I prefer) and 2705 (a recommended version). They are both working fine.

**Pull request merged**